### ISSUE TITLE
Offline Routing: Display exit number

### ISSUE THREAD

**Issue by OpenSourceForEver** [osmandapp/Osmand#1347]
Hi. When you take an exit in a highway Osmand do not say take exit 34, osmand only says turn right.

When you take an exit in a roundabout Osmand says destination info only when there is not name  or reference in the way. If there is name or reference in the way Osmand only says take first exit without destination info.

Is it a problem with spanish voice?

**Comment by jfburke3** [comment]
No, it does the same thing in English.  There's another thread on this somewhere, it might be in the Google group instead of here.  I'll try to find it.

**Comment by Norbert80** [comment]
Same with german.

**Comment by OpenSourceForEver** [comment]
Is there any solution for it?

**Comment by vshcherb** [comment]
So far not. This data is not even present for routing. A couple preliminary steps need to be done. I think it is a duplicate issue, will search through existing Offline Routing issues and attach it.

**Comment by OpenSourceForEver** [comment]
Do you mean I have to create an issue in offline routing? I don't find any similar issue. If solving it is very difficult i won't attach it.
Thanks.

**Comment by OpenSourceForEver** [comment]
Sorry, I didn't know Who had to attach it in Offline routing. I am novice here. I just want to help.

Thank you

**Comment by vshcherb** [comment]
It's fine, this is my work.

**Comment by 9662** [comment]
Question: is the bug about expecting OsmAnd to report the exit number *if that information is available on OpenStreetMap*? In my experience, not many exits are as well tagged as they could be in OSM (on days when the OSRM website works, I click on "highlight unnamed roads" and tag them).

**Comment by zeromap** [comment]
@vshcherb Would you be able to expand on some of the challenges you are facing implementing this feature? In the US at least, data quality is quite good. Or are the issues with the tags themselves? As an OSM contributor, I would like to help in any way possible.

**Comment by vshcherb** [comment]
The features is actually implemented, the only thing is missing UI. I think I just get stuck where and how to display it.

**Comment by OpenSourceForEver** [comment]
Maybe a icon like these at the same location that another icons are at the moment.

![1](https://cloud.githubusercontent.com/assets/12786225/11770384/e72d50e2-a1fc-11e5-9eb2-2942728e0715.gif)

![2](https://cloud.githubusercontent.com/assets/12786225/11770386/f097e23c-a1fc-11e5-8f58-ea986fd21e81.jpg)

**Comment by OpenSourceForEver** [comment]
When there is a destination tag without a streetname tag, Osmand says: "Turn left to (destination)". It is ok

When there is a destination tag with a streetname tag, Osmand says: "Turn right on (streetname) (destination)". I think osmand could says: "Turn right on (streetname) to (destination)"

What do you think?

**Comment by vshcherb** [comment]
I just added that information to route description (visible on show info route screen) https://github.com/osmandapp/Osmand/commit/c07b8eb4051d200fe4883932208297141d95541a#diff-3309c456b457277799e09eed648d65fcR264
But I totally don't envision how to add to UI right now.

At this point, http://www.openstreetmap.org/way/75928702
we have name,  destination, destination:ref + motorway_junction:ref + motorway_junction:name that's too much. 
That what I mean we miss proper design that would make features compatible and also explain how to display them.

**Comment by zeromap** [comment]
>that's too much. 

Adding support for motorway_junction ref alone would be an improvement. Would it make sense to interpret tags differently based on selected driving region in the general settings?

**Comment by vshcherb** [comment]
I think we need to think about clear rules what to display in what country first, before we start implementation and also we need to double check if there is a mapping tradition already in that region.

**Comment by zeromap** [comment]
I only have direct knowledge of how things are in the United States, and consensus is growing around [this method](https://wiki.openstreetmap.org/wiki/Exit_Info).

**Comment by OpenSourceForEver** [comment]
I just added that information to route description (visible on show info route screen) c07b8eb#diff-3309c456b457277799e09eed648d65fcR264

Info route screen do not show "take exit xx on xxx" yet. It only shows exit number with a number at the end of the text.

![screenshot_2016-01-09-01-30-56](https://cloud.githubusercontent.com/assets/12786225/12213046/16c1a194-b672-11e5-9220-4d156b4bc0bd.png)

**Comment by caederus-osmand** [comment]
I see things have improved since the original issue report, but I still find the directions for highway junctions quite confusing (this is in the UK, where the OSM mark-up is generally good).  For a specific example, leaving the M4 motorway towards the M32 motorway at this node: http://www.openstreetmap.org/node/309384

a) The "Route details" text reads: "Keep left and go 19 M32 Bristol", which isn't bad, but it would be improved by the additon of the word "exit" or "junction" before  the description of the node, i.e. "Keep left and go junction 19 M32 Bristol" - the justification being that the node being described is of highway type "motorway_junction", and the extra word is a clear indication that one should expect to leave the main carriageway.

b) The spoken direction is: "Keep left on M4", which is very confusing indeed, since we are supposed to be leaving the M4, not staying on it!  Is there any good reason for this to mention the current road, rather than the destination, as in the route details?

**Comment by polarbearing** [comment]
I would appreciate some progress with the exit numbers. Here is a mockup of an idea where to show it. And it would be good to include that in the TTS in the early turn announcements. 
As said before, for the route description, the exit name and number hangs at the end of the instruction without being labelled, it should read e.g. "Exit: number name"

![mockup_exit_number](https://cloud.githubusercontent.com/assets/8266355/26552124/18ab8b7a-4486-11e7-8a09-2bb3e368c2a8.png)

**Comment by osmvermesser** [comment]
+1 to THW image by polarbearing

Jan

**Comment by alarije** [comment]
It's important that that exit number (ref) are shown in the screen and spoken in navigation. 

Now it says "keep left", is there is only one exit it's acceptable, but at ring roads in Spain there sometimes multiple exits at almost the same place, ex. exit 17A / 17B / 17C are at the same km, so you don't know what exit to take.

Ex. https://www.google.es/maps/@37.377757,-6.0266204,3a,75y,188.35h,96.58t/data=!3m6!1e1!3m4!1sjReDgtVnQ2luBYyZRrLS7w!2e0!7i13312!8i6656

https://www.google.es/maps/@37.3890942,-6.0292287,3a,31.4y,120.23h,92.48t/data=!3m6!1e1!3m4!1sBNUpA24CnxxCLZc5r4RlEQ!2e0!7i13312!8i6656

**Comment by meoberto** [comment]
Navit 0.5.0 already had this implemented, using the generic motorway exit icon to signal that the following turn to take is a motorway junction.

The tag "highway=motorway_junction" applies, so this is data that is already on OSM. It would be a matter of showing that icon on the navigation instructions, then probably the ref beside it.

Shame this didn't make it into 3.1 though, I'm going on a trip soon-ish and this would be a really nice feature to have on a navigation software.

**Comment by osmvermesser** [comment]
7 month ago "vshcherb" write this feature will come in Version 3.1!

now there is 3.1.4 current and this ticket open.

I think it is time to integrate this feature - even by the gold Argumente in the post before!!

Jan

**Comment by ArenaL5** [comment]
There are a lot of duplicates of this issue, and, even though most of them are closed, it's confusing to follow the discussion. A couple of comments there are worth copypasting:

@wehkah, #1451 
>What about announcements like "Take the next exit and follow the signs towards X", or "At the next exit, >follow the signs towards Y", or similar?
>
>That way you don't have to worry about different regional exceptions and you also don't have to worry >about diversions because of construction sites, crashes or other disturbances that lead to closed roads, >because there will be signs or traffic wardens telling drivers where to go.

@martenjj, #6875 
>The instruction could say something like "Turn off in DISTANCE at junction REF NAME" or "Leave at >junction REF NAME".
>
>Leaving the motorway could be detected by either:
>(a) passing over a node of type highway=motorway_junction (with name= the junction/destination name >and ref= the junction number);
>(b) if the junction is not tagged properly as above, a transition between highway=motorway and >highway=motorway_link (also trunk => trunk_link and primary -> primary_link).

I prefer martenjj's approach (using “the next exit” if the junction has no name or number), as it's closer to what we have now. Whichever we use, I think we should only use `ref` in the node if there is no `junction:ref` in the link.

(It's the reason why `junction:ref` was created, as far as I know. It's useful when two exits share nodes, and it can be used in any type of road, not only motorways.)

**Comment by osmvermesser** [comment]
When you show a symbol like tunnel for the exit number dont forget that sometimes exit in or at the end of tunnels. The symbols should notify overlap.

**Comment by vshcherb** [comment]
Implemented in 3.6

**Comment by osmvermesser** [comment]
The new function is very helpful. Only the icon leads to confusion. The Ref symbol for other street classes is also displayed in the same place.

The blue symbol looks confusingly similar to the German motorway symbol.

It would be desirable to exchange this for a clear symbol.


An alternative would be the icon with an arrow pointing upwards to the right or a white circle around the exit number.

Before I write this I have asked in the German OSM forum and there is a common opinion that this icon is unsuitable.

https://forum.openstreetmap.org/viewtopic.php?id=68830

**Comment by dead10ck** [comment]
I'm not sure if this feature was regressed or what, but I do not see exit numbers announced. See #8767 for details.

**Comment by sonora** [comment]
I can in deed reproduce that -while the voice prompts work as intended- the on-screen display (top of the map screen) **fails to display for exits which carry only a ref tag** and not at least an additional name or destination tag (in which case it displays ok).

I believe this to be no regression, but a glitch of the original implementation in January.

**Comment by dead10ck** [comment]
It's worth noting that, per the wiki, the recommended way to map and tag motorway exits is to have the junction node tagged with the ref number and name (if the **junction itself** has a name, which most do not in the US), and the motorway link which branches from the motorway junction node tagged with the junction ref and the destination.

So the common case in the US for a motorway exit will be to have a junction node with only a `ref` tag, from which branches a motorway link way which has a `junction:ref` and `destination` or `destination:street` tags. In this case, nothing displays.

**Comment by sonora** [comment]
PS: In addition to the remaining issue pinpointed in my post above, the [OSM wiki](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dmotorway_junction) indicates that, while `ref` and `name` of an exit should be tagged on the the `motorway_junction` itself, the `destination=*` tag should be placed '_on the way that branches off from the node motorway_junction_'.

We may infer from that that for exits it would make sense for us to 'pre-read' the `destination` from the subsequent road segment?

**Comment by dead10ck** [comment]
@sonora that seems necessary given the recommended mapping schema. The destination tag is on the following way, so that's where it needs to get the information.

**Comment by vshcherb** [comment]
The issue was implemented in 3.6 but there was an issue for certain type of roads that whole address bar disappears, it is already fixed in 3.7

There is no sense for reopening that issue cause it is already release as is in 3.6.

**Comment by vshcherb** [comment]
Please open a separate issue with map details

**Comment by dead10ck** [comment]
@vshcherb oh, that's good to hear! It sounded like from what @sonora said that there might still be an issue with how it fetches the street name to display, but I suppose this issue was particularly about the exit number, so maybe #8767 is the place to address that.

**Comment by vshcherb** [comment]
It still looks quite overwhelmed with information to me, so I will wait @sonora to summarize it.   don't know which part of street is better to display from following road or from point itself cause it could vary quite a lot in the OSM data.

**Comment by sonora** [comment]
We're good now, I think! See my comments and screenshot taken with the currently nightly in #8767.
